pkg/mapi2capi/aws.go

```diff
@@ -0,0 +1,431 @@
+
+	capiAdditionalSecurityGroups := convertAWSSecurityGroupstoCAPI(p.Spec.SecurityGroups)
+
+	capiAdditionalAWSTags := convertAWSTagsToCAPI(p.Spec.Tags)
+
+	capiIAMInstanceProfile := convertIAMInstanceProfiletoCAPI(p.Spec.IAMInstanceProfile)
+
+	capiMetadataServiceOptions := convertMetadataServiceOptionstoCAPI(p.Spec.MetadataServiceOptions)
+
+	capiSpotMarketOptions := convertAWSSpotMarketOptionsToCAPI(p.Spec.SpotMarketOptions)
+
+	capiAWSResourceReference := convertAWSResourceReferenceToCAPI(p.Spec.Subnet)
+
+	capiAWSAMIReference, err := convertAWSAMIResourceReferenceToCAPI(p.Spec.AMI)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	spec := capav1.AWSMachineTemplateSpec{
+		Template: capav1.AWSMachineTemplateResource{
+			Spec: capav1.AWSMachineSpec{
+				AMI:                      capiAWSAMIReference,
+				AdditionalSecurityGroups: capiAdditionalSecurityGroups,
+				AdditionalTags:           capiAdditionalAWSTags,
+				// CloudInit. Not defined as we use ignition in OpenShift.
+				IAMInstanceProfile: capiIAMInstanceProfile,
+				Ignition: &capav1.Ignition{
+					Version:     "3.4",                                               // Hardcoded for OpenShift.
+					StorageType: capav1.IgnitionStorageTypeOptionUnencryptedUserData, // Hardcoded for OpenShift.
+				},
+				// ImageLookupBaseOS. Not used in OpenShift.
+				// ImageLookupFormat. Not used in OpenShift.
+				// ImageLookupOrg. Not used in OpenShift.
+
+				// TODO: what to do with instanceID? in MAPI that's not in AWSMachineProviderConfig but outside of it.
+				// Is this propagated down by the CAPA controller automatically from the CAPI Machine
+
+				// InstanceID. This is dynamically populated by the controller.
+
+				InstanceMetadataOptions: capiMetadataServiceOptions,
+				InstanceType:            p.Spec.InstanceType,
+				// NetworkInterfaces. Not used in OpenShift.
+				NonRootVolumes:     nonRootVolumes,
+				PlacementGroupName: p.Spec.PlacementGroupName,
+
+				// TODO: what to do with providerID? in MAPI that's not in AWSMachineProviderConfig but outside of it.
+				// Is this propagated down by the CAPA controller automatically from the CAPI Machine
+				// ProviderID. This is dynamically populated by the controller.
+				PublicIP:             p.Spec.PublicIP,
+				RootVolume:           rootVolume,
+				SSHKeyName:           p.Spec.KeyName,
+				SpotMarketOptions:    capiSpotMarketOptions,
+				Subnet:               capiAWSResourceReference,
+				Tenancy:              string(p.Spec.Placement.Tenancy),
+				UncompressedUserData: ptr.To(true),
+			},
+		},
+	}
+
+	if len(errs) > 0 {
+		return capav1.AWSMachineTemplateSpec{}, warnings, utilerrors.NewAggregate(errs)
+	}
+
+	return spec, warnings, nil
+}
+
+// AWSProviderSpecFromRawExtension unmarshals a raw extension into an AWSMachineProviderSpec type.
+func AWSProviderSpecFromRawExtension(rawExtension *runtime.RawExtension) (mapiv1.AWSMachineProviderConfig, error) {
```

Comment:
Does this need to be public?

Reply:
I think this is useful also externally if one wants to extract the providerSpec from a MAPI MachineSet to then only feed that in the converter.

Reply:
Ack, I think maybe lets move all of these out to a utils package somewhere, need to make sure they are actually used in more than one place though